### PR TITLE
C++: Reduce number of FPs `cpp/guarded-free` and turn `if(x) { free(x) }` cases from FNs to TPs

### PR DESCRIPTION
**Pull request by jketema**
The additional conditions to reduce FPs are maybe somewhat on the strict side, but given that a MRVA 1000 experiment already gives over 5k results, I don't think that this much of a problem. As far as I can tell there is just one remaining FP in MRVA (which occurs a few times), and which comes from the libpng library.

https://github.com/naturalatlas/node-gdal/blob/c83e7858a9ec566cc91d65db74fd07b99789c0f0/deps/libgdal/gdal/frmts/png/libpng/pngmem.c#L541-L560

Given that this is now the only FP, I did not investigate this further. I think we could promote the query with these changes from this PR.

#### All query authors

- [ ] A change note is added if necessary. See [the documentation](https://github.com/github/codeql/blob/main/docs/change-notes.md) in this repository.
- [ ] All new queries have appropriate `.qhelp`. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-help-style-guide.md) in this repository.
- [x] QL tests are added if necessary. See [Testing custom queries](https://docs.github.com/en/code-security/codeql-cli/using-the-advanced-functionality-of-the-codeql-cli/testing-custom-queries) in the GitHub documentation.
- [x] New and changed queries have correct query metadata. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-metadata-style-guide.md) in this repository.

#### Internal query authors only

- [ ] Autofixes generated based on these changes are valid, only needed if this PR makes significant changes to `.ql`, `.qll`, or `.qhelp` files. See [the documentation](https://github.com/github/codeql-team/blob/main/docs/best-practices/validating-autofix-for-query-changes.md) (internal access required).
- [ ] Changes are validated [at scale](https://github.com/github/codeql-dca/) (internal access required).
- [ ] Adding a new query? Consider also [adding the query to autofix](https://github.com/github/codeml-autofix/blob/main/docs/updating-query-support.md#adding-a-new-query-to-the-query-suite).

### CODE REVIEW

File: cpp/ql/src/experimental/Best Practices/GuardedFree.ql [hunk 1]
```diff
@@ -18,9 +18,28 @@ class FreeCall extends FunctionCall {
...
   fc.getArgument(0) = v.getAnAccess() and
-  bb = fc.getEnclosingStmt()
+  bb = fc.getBasicBlock() and
+  (
+    bb = fc.getEnclosingStmt()
+    or
+    strictcount(bb.(BlockStmt).getAStmt()) = 1
+  ) and
+  strictcount(BasicBlock bb2 | gc.ensuresEq(_, 0, bb2, _) | bb2) = 1 and
+  not fc.isInMacroExpansion() and
```
Comment (geoffw0) [inline cpp/ql/src/experimental/Best Practices/GuardedFree.ql:41]: I don't particularly like this exclusion, but I can see that it circumvents some really difficult cases and helps get us a good set of results.

**Review by geoffw0 — COMMENTED** [review]
Couple of small requests, otherwise LGTM.

Also deserves a DCA run before merging.  I gather you have reviewed MRVA differences already.

File: cpp/ql/src/experimental/Best Practices/GuardedFree.ql [hunk 2]
```diff
@@ -18,9 +18,28 @@ class FreeCall extends FunctionCall {
   FreeCall() { this.getTarget().hasGlobalName("free") }
 }
 
+predicate blockContainsPreprocessorBranches(BasicBlock bb) {
+  exists(PreprocessorBranch ppb, Location bbLoc, Location ppbLoc |
+    bbLoc = bb.(Stmt).getLocation() and ppbLoc = ppb.getLocation()
+  |
+    bbLoc.getStartLine() < ppbLoc.getStartLine() and
+    ppbLoc.getEndLine() < bbLoc.getEndLine()
```
[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/src/experimental/Best Practices/GuardedFree.ql:27]: Thanks for spotting that. Done.

File: cpp/ql/src/experimental/Best Practices/GuardedFree.ql [hunk 3]
```diff
@@ -18,9 +18,28 @@ class FreeCall extends FunctionCall {
...
 from GuardCondition gc, FreeCall fc, Variable v, BasicBlock bb
 where
   gc.ensuresEq(v.getAnAccess(), 0, bb, false) and
   fc.getArgument(0) = v.getAnAccess() and
-  bb = fc.getEnclosingStmt()
+  bb = fc.getBasicBlock() and
+  (
+    bb = fc.getEnclosingStmt()
+    or
+    strictcount(bb.(BlockStmt).getAStmt()) = 1
```
[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/src/experimental/Best Practices/GuardedFree.ql:41]: Done.

File: cpp/ql/src/experimental/Best Practices/GuardedFree.ql (hunk 1 above)
Comment (jketema) [reply cpp/ql/src/experimental/Best Practices/GuardedFree.ql:44]: Agreed.

**Comment by jketema** [comment]
> Also deserves a DCA run before merging.

How do you propose I run DCA? Note that this is still an experimental query, so it won't be run by default.

>I gather you have reviewed MRVA differences already.

Correct. Checking for the file when in the proprocessor block case, might deserve some further investigation though.

**Comment by geoffw0** [comment]
> How do you propose I run DCA? Note that this is still an experimental query, so it won't be run by default.

Ah, perhaps we can save this until the PR that promotes the query then.  FWIW I think it is possible to specify a single query rather than a suite when running DCA - if so, that might work around the problem, though I guess nothing would be cached so the performance data might still be a bit off.

> Checking for the file when in the proprocessor block case, might deserve some further investigation though.

Yeah, this bug could have been almost randomly hiding results - good or bad.

**Review by geoffw0 — APPROVED** [review]
:+1:

Approved, on the assumption that you briefly re-check some results on MRVA following the fix to `blockContainsPreprocessorBranches`.

**Comment by jketema** [comment]
> FWIW I think it is possible to specify a single query rather than a suite when running DCA 

Good point. Let me run that anyway, so we can at least see if there are any alert differences, and it doesn't blow up completely analysis time-wise.

**Comment by geoffw0** [comment]
DCA shows 99 added, 2866 removed results for `cpp/guarded-free`.  Unfortunately we don't seem to be able to see the individual changes, but I suppose we were expecting some new results for the `{ free(x); }` pattern.  Analysis time looks to be unaffected. :+1:

**Comment by jketema** [comment]
> Approved, on the assumption that you briefly re-check some results on MRVA following the fix to `blockContainsPreprocessorBranches`.

There are about 1.6k new results. They all look genuine.

> DCA shows 99 added, 2866 removed results for `cpp/guarded-free`. Unfortunately we don't seem to be able to see the individual changes, but I suppose we were expecting some new results for the `{ free(x); }` pattern. Analysis time looks to be unaffected. 👍

Yup. I've asked internally whether I can still somehow generate the table.

**Comment by jketema** [comment]
I spot checked some of the alerts that disappeared in DCA. Most of them seem to be due to
```cpp
 strictcount(BasicBlock bb2 | gc.ensuresEq(_, 0, bb2, _) | bb2) = 1
```
And it's correct that they disappeared, because the code depends on the guard being there.

**Pull request merged**